### ISSUE TITLE
`make latex` able to rebuild only what is needed, just as `make html`?

### ISSUE THREAD

**Issue by dbitouze** [sphinx-doc/sphinx#13181]
**Is your feature request related to a problem? Please describe.**
I'm always frustrated when I have to recompile the whole PDF of my project, even though only one of its many source files has changed, whereas, for HTML output, only the corresponding HTML file is rebuilt.

**Describe the solution you'd like**
It would be nice to have `make latex` as `make html` able to rebuild only what is needed.  For this, make latex would have to copy the HTML builder but it looks needed then for its mechanism to produce in build repertory as many `.tex` files as there are source files. Hopefully it should be possible from one extra "main" file to "input" all the small "unit files".  Is this possible?

**Comment by dbitouze** [comment]
@jfbu "Joining" all small PDF's will make all of them start a new page and at the top of the page, which isn't what is expected, is it?

IMHO, obtaining only a small PDF would be just for checking the layout of the modifications just done in a single source file.

**Comment by electric-coder** [comment]
>  able to rebuild only what is needed

The way HTML is organized part of the build is split across several files that are linked (e.g. rebuilding a single page assumes no significant changes happened to the index) but PDF is different in that it's compiled so to rebuild one part you have to rebuild the whole think, compile it and finally compress it.

**Comment by electric-coder** [comment]
@dbitouze LaTeX isn't my *forte* but I don't think that having *"individual PDFs"* would speed up the build of the final monolith due to requirements on internal consistency that HTML does not impose but the other formats do. So, if there's an omission in the index both PDF and LaTeX are likely to throw errors wheres a split HTML won't. 

Your argument is somewhat conceptual but, as @jfbu hinted, that's just not how the other document's build process works because there are requirements on internal consistency, it seems LaTeX itself doesn't provide the necessary underlying tooling; see [Does LaTeX have to reprocess included files that haven't changed?][1]

[1]: https://tex.stackexchange.com/a/55610

**Comment by dbitouze** [comment]
@electric-coder 

> LaTeX isn't my _forte_ but I don't think that having _"individual PDFs"_ would speed up the build of the final monolith [...]

But that's not what I'm about :) My aim is not to speed up the construction of the final monolith, just to be able to construct a partial PDF containing only the few pages corresponding to the modified source file(s).

For example, our [LaTeX FAQ](https://faq.gutenberg-asso.fr/) is made up of almost 1400 source files (`.md`) and [the complete PDF](https://www.gutenberg-asso.fr/IMG/pdf/faqlatexgutenberg.pdf.pdf) is over 2050 pages long. When a contributor modifies something in one of the 1400 source files (`.md`), for example the one corresponding to [this short FAQ](https://faq.gutenberg-asso.fr/4_domaines_specialises/mathematiques/equations/numerotation_d_equations/changer_la_numerotation_des_equations.html), he would be interested in checking the PDF output of the modified file, which is quickly constructed because it contains only (or almost only) the one or two corresponding pages.

**Comment by dbitouze** [comment]
@jfbu 

> I see one major difficulty regarding cross references, when working with `\input`. As @electric-coder LaTeX is not so much my _forte_ (I am more at ease with Plain TeX), but I am vaguely aware of `\include/\includeonly`, which could help with the cross-references, although I have not practiced that for a long time, and I remember `\include` does a clearpage which causes a problem if we are to structure the main .tex file as a bunch of `\include` with respect to the "atomic" ones.

You're right about `\input` which would miss the cross references if targets are in files not 
“`\input`ed” in order to speed up the compilation of the partial PDF. And, indeed, `\include`/`\includeonly` is better in this respect, as long as the whole PDF has been built (and stabilized) at least once. But:

- missing the cross-references (replaced by “**??**” by LaTeX) is most of the time acceptable if the purpose is just to check the PDF output,
- indeed, `\include` does a `\clearpage` but it could be redefined into `\input` for the whole PDF.

**Comment by electric-coder** [comment]
> just to be able to construct a partial PDF containing only the few pages corresponding to the modified source file(s).

Lets split this sentence:

> 1. just to be able to construct a partial PDF containing only the few pages

The two common alternatives:

A. Construct a second `/doc` directory to build a separate documentation containing links to only the sources you want to build - that is: a second main index `toctree` changed manually to only include the files that changed in your case. (One project using this kind of layout is [Pyramid docs][1] their motivation being to have one *"organic"* documentation and another inner [API doc with linear package/module layout for machine consumption][2]). You'll be running into the difficulty: [Can sphinx link to documents that are not located in directories below the root document?][4].

B. Same as the above but keeping a "development" version of your docs separately from the "production" version where you selectively wrap whatever is causing delays (the main toctree) in an [`ifconfig`][5] controlled via a `conf.py` option.

> 2. the few pages corresponding to the modified source file(s)

Here Sphinx itself keeps track of changed files (the manual solutions above being obviously inconvenient). It's a lot easier to just use the alternative proposed by @jfbu 

> checking the HTML build is a better alternative

---

> **Describe the solution you'd like**
>
> It would be nice to have `make latex` as `make html` able to rebuild only what is needed.

So a development convenience tool (as a builder CLI flag) to build just the the latest changed sources as individual PDFs ignoring any internal inconsistencies arising from incompleteness (a kind of [`--keep-going`][3]) ... There's definitely a valid use case here, there's also a promising change with #12882 to speed up builds by an order of magnitude that might make this kind of issue with large document a non-problem.

[1]: https://github.com/Pylons/pyramid/tree/main/docs
[2]: https://github.com/Pylons/pyramid/tree/main/docs/api
[3]: https://www.sphinx-doc.org/en/master/man/sphinx-build.html#cmdoption-sphinx-build-keep-going
[4]: https://stackoverflow.com/q/10199233
[5]: https://www.sphinx-doc.org/en/master/usage/extensions/ifconfig.html
